### PR TITLE
Allow access to UDTF in `SessionContext`

### LINKED ISSUE #9203: add `udtf` to `FunctionRegistry
### Is your feature request related to a problem or challenge?

currently only the session context has `udtf` function. I think that this should instead be part of the `FunctionRegistry` like the rest of the `udf` variants

### Describe the solution you'd like

add `register_udtf` and `udtf` to `FunctionRegistry`

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by linhr**
## Which issue does this PR close?

Partially closes #9203.

## Rationale for this change

The `FunctionRegistry` trait does not include methods to manage UDTF. This is because adding the methods would cause circular dependency on `TableProvider`, as explained in #9203. However, it would be useful if the user can still access/deregister the table functions they have registered. We already have the `register_udtf()` method for `SessionState` and `SessionContext`, though they are not part of the `impl FunctionRegistry` block.

## What changes are included in this PR?

This PR adds public API for `SessionContext` and `SessionState` so that the user can look up and deregister UDTF. The method is not part of the `FunctionRegistry` trait though.

## Are these changes tested?

Yes. A unit test is added for the changes.

## Are there any user-facing changes?

Yes, this PR contains backward-compatible changes to the public API.

### CODE REVIEW

File: datafusion/core/src/execution/context/mod.rs [hunk 1]
```diff
@@ -1277,6 +1297,11 @@ impl SessionContext {
         state
     }
 
+    /// Get reference to [`SessionState`]
+    pub fn state_ref(&self) -> Arc<RwLock<SessionState>> {
```
Comment (linhr) [inline datafusion/core/src/execution/context/mod.rs:1301]: This is not directly related to the issue we are resolving, but I found it useful for the user to get the `SessionState` reference. So this method is added as a complement to `state()` and `state_weak_ref()` that we already have.

**Review by alamb — APPROVED** [review]
Thank you @linhr  -- this is a very nice first contribution and a nice improvement 🙏 

I have started CI and once that passes I think this PR is ready to go

**Comment by alamb** [comment]
Thanks again @linhr  and @shehabgamin
